server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -4852,9 +4860,58 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
                 _tmplService.attachIso(tmpl.getId(), vm.getId());
             }
         }
+
+        // Add extraConfig to user_vm_details table
+        Account caller = CallContext.current().getCallingAccount();
+        String extraConfig = cmd.getExtraConfig();
+        Long vmId = vm.getId();
+        if (!Strings.isNullOrEmpty(extraConfig)) {
+            addExtraConfig(vm, caller, extraConfig);
+        }
+
         return vm;
     }
 
+    protected void addExtraConfig(UserVm vm, Account caller, String extraConfig) {
+        Long callerId = caller.getId();
+        String decodedUrl = decodeExtraConfig(extraConfig);
+        validateConfigOption(decodedUrl, callerId);
+        if (EnableAdditionalVmConfig.valueIn(callerId)) {
+             if (vm.getHypervisorType() == HypervisorType.VMware) {
+                 String[] configDataArr = decodedUrl.split("\\r?\\n");
+                 for (String config: configDataArr) {
+                     String[] keyValue = config.split("=");
+                     try {
+                         userVmDetailsDao.addDetail(vm.getId(), keyValue[0], keyValue[1], true);
+                     } catch (ArrayIndexOutOfBoundsException e) {
+                         throw new CloudRuntimeException("Issue occurred during parsing of:" + config);
+                     }
+                 }
+             } else {
+                 userVmDetailsDao.addDetail(vm.getId(), ApiConstants.EXTRA_CONFIG, decodedUrl, true);
+             }
+        }
+    }
+
+    protected void validateConfigOption(String decodedUrl, Long callerId) {
+        String[] accessConfigBlackList = AdditionalConfigBlackList.valueIn(callerId).split(",");
+        for (String configOption : accessConfigBlackList) {
+            if (decodedUrl.contains(configOption)) {
```

Comment:
I think at this place you should also split the decoded url and check if the any key matches the blacklist element (not contains). For example these fake scenarios will throw an exception as it is right now:

blacklist = cpu,mem
extraconfig: "key1=allowcpu"

blacklist = cpu,mem
extraconfig: "memory=1111"

Reply:
Implemented separate parsing mechanisms per hypervisor.
kvm -> Check XML tags against blacklist
Vmware/Xen -> Check keys against blacklist